File: agent/src/main/java/pt/um/lei/masb/agent/ReceiveMessages.java
package pt.um.lei.masb.agent;

import jade.content.ContentElement;
import jade.content.lang.Codec;
import jade.content.lang.sl.SLCodec;
import jade.content.onto.BeanOntologyException;
import jade.content.onto.Ontology;
import jade.content.onto.OntologyException;
import jade.core.behaviours.Behaviour;
import jade.lang.acl.ACLMessage;
import jade.lang.acl.MessageTemplate;
import pt.um.lei.masb.agent.data.block.BlockOntology;
import pt.um.lei.masb.agent.data.transaction.TransactionOntology;
import pt.um.lei.masb.blockchain.Block;
import pt.um.lei.masb.blockchain.BlockChain;
import pt.um.lei.masb.blockchain.Transaction;

public class ReceiveMessages extends Behaviour {
    private Codec codec = new SLCodec();
    private Ontology txOntology,blOntology;
    private Transaction tx;
    private Block bl;
    private BlockChain bc;

    public ReceiveMessages (BlockChain bc)  {
        this.bc=bc;
        try {
            txOntology=new TransactionOntology();
            blOntology=new BlockOntology();
        } catch (BeanOntologyException e) {
            e.printStackTrace();
        }
    }


    @Override
    public void action() {
        var mt = MessageTemplate.and(MessageTemplate.MatchLanguage(codec.getName()),
                                     MessageTemplate.MatchOntology(txOntology.getName()) );


        var txmsg = myAgent.receive(mt);
        try {
            if (txmsg!=null) {
                ContentElement txce = myAgent.getContentManager().extractContent(txmsg);
                tx = (Transaction) txce;
                bc.getLastBlock().addTransaction(tx);
            }
        } catch (Codec.CodecException | OntologyException e) {
            e.printStackTrace();
        }
        var blocksReq= myAgent.receive();
        if (blocksReq!=null){
            var rHeight=Long.parseLong(blocksReq.getContent());
            //Send number of missing blocks
            var sendMissingNum = new ACLMessage(ACLMessage.INFORM);
            var missingNum = bc.getLastBlock().getHeader().getBlockheight() - rHeight;
            sendMissingNum.setContent(String.valueOf(missingNum<=0?0:missingNum));

            //Send missing blocks
            while (missingNum>0){
                var codec = new SLCodec();
                var blmsg = new ACLMessage(ACLMessage.INFORM);
                //Block of the blockchain is not JADE serializable
                //Need to convert into JBlock of block ontology, in block.ontology package
                //in order to actually send it.
                myAgent.getContentManager().fillContent(blmsg,bc.getBlockByHeight(rHeight));
                blmsg.addReceiver(blocksReq.getSender());
                blmsg.setLanguage(codec.getName());
                blmsg.setOntology(blOntology.getName());
                rHeight++;
                missingNum--;
            }
        }
    }

    @Override
    public boolean done() {
        return false;
    }
}

File: agent/src/main/kotlin/pt/um/lei/masb/agent/behaviours/ReceiveMessages.kt
package pt.um.lei.masb.agent.behaviours

import jade.content.lang.sl.SLCodec
import jade.core.behaviours.Behaviour
import jade.lang.acl.ACLMessage
import jade.lang.acl.MessageTemplate
import mu.KLogging
import pt.um.lei.masb.agent.data.AgentPeers
import pt.um.lei.masb.agent.data.convertToJadeBlock
import pt.um.lei.masb.agent.messaging.block.BlockOntology
import pt.um.lei.masb.agent.messaging.block.ontology.actions.DiffuseBlock
import pt.um.lei.masb.agent.messaging.transaction.TransactionOntology
import pt.um.lei.masb.blockchain.data.BlockChainData
import pt.um.lei.masb.blockchain.ledger.Block
import pt.um.lei.masb.blockchain.ledger.Transaction
import pt.um.lei.masb.blockchain.service.ChainHandle
import pt.um.lei.masb.blockchain.service.results.LoadResult

/**
 * Behaviour for handling incoming messages related to Transactions and Blocks.
 * TODO: Split into behaviours for each.
 */
class ReceiveMessages(
    private val sc: ChainHandle,
    private val agentPeers: AgentPeers,
    private val clazz: Class<out BlockChainData>
//    private val srl: SerializationStrategy<T>
) : Behaviour() {
    private val codec = SLCodec()
    private lateinit var tx: Transaction
    private lateinit var bl: Block

    @Suppress("UNCHECKED_CAST")
    override fun action() {
        val mt = MessageTemplate.and(
            MessageTemplate.MatchLanguage(codec.name),
            MessageTemplate.MatchOntology(TransactionOntology.name)
        )


        /**
        val txmsg = myAgent.receive(mt)
        try {
        if (txmsg != null) {
        val txce = myAgent.contentManager.extractContent(txmsg)
        //TODO: Must extract a transaction from a content element.
        tx = txce as Transaction
        sc.lastBlock?.addTransaction(tx)
        }
        } catch (e: Codec.CodecException) {
        logger.error(e){}
        } catch (e: OntologyException) {
        logger.error(e){}
        }
         */

        val bt = MessageTemplate.and(
            MessageTemplate.and(
                MessageTemplate.MatchLanguage(codec.name),
                MessageTemplate.MatchOntology(BlockOntology.name)
            ),
            MessageTemplate.and(
                MessageTemplate.MatchProtocol("blockNotice"),
                MessageTemplate.MatchPerformative(ACLMessage.REQUEST)
            )
        )


        val blocksReq = myAgent.receive()
        if (blocksReq != null) {
            var rHeight = java.lang.Long.parseLong(blocksReq.content)
            //Send number of missing blocks
            val sendMissingNum = blocksReq.createReply()
            var missingNum = sc.lastBlockHeader.let {
                when (it) {
                    is LoadResult.Success -> it.data.blockheight - rHeight
                    else -> -1
                }
            }
            sendMissingNum.content = (if (missingNum < 0) -1 else missingNum).toString()
            agent.send(sendMissingNum)

            //Send missing blocks
            loop@ while (missingNum > 0) {
                val codec = SLCodec()
                val blmsg = ACLMessage(ACLMessage.INFORM)
                val blk = sc.getBlockByHeight(rHeight)
                when (blk) {
                    is LoadResult.Success -> {
                        myAgent.contentManager
                            .fillContent(
                                blmsg,
                                DiffuseBlock(
                                    convertToJadeBlock(blk.data, clazz),
                                    blocksReq.sender
                                )
                            )
                        blmsg.addReceiver(blocksReq.sender)
                        blmsg.language = codec.name
                        blmsg.ontology = BlockOntology.name
                        rHeight++
                        missingNum--
                    }
                    else -> {
                        logger.error {
                            "Nonexistent block referenced at height: $rHeight"
                        }
                        break@loop
                    }
                }
            }
        }
    }

    override fun done(): Boolean {
        return false
    }

    companion object : KLogging()
}
